src/app/ReadClient.cpp

```diff
@@ -77,95 +78,112 @@ void ReadClient::MoveToState(const ClientState aTargetState)
-                                       size_t aEventPathParamsListSize, AttributePathParams * apAttributePathParamsList,
-                                       size_t aAttributePathParamsListSize, EventNumber aEventNumber)
+CHIP_ERROR ReadClient::GenerateReadRequest(System::PacketBufferTLVWriter & aWriter, EventPathParams * apEventPathParamsList,
+                                           size_t aEventPathParamsListSize, AttributePathParams * apAttributePathParamsList,
+                                           size_t aAttributePathParamsListSize, EventNumber aEventNumber)
 {
-    // TODO: SendRequest parameter is too long, need to have the structure to represent it
     CHIP_ERROR err = CHIP_NO_ERROR;
-    System::PacketBufferHandle msgBuf;
-    ChipLogDetail(DataManagement, "%s: Client[%u] [%5.5s]", __func__,
-                  InteractionModelEngine::GetInstance()->GetReadClientArrayIndex(this), GetStateStr());
-    VerifyOrExit(ClientState::Initialized == mState, err = CHIP_ERROR_INCORRECT_STATE);
-    VerifyOrExit(mpDelegate != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
-
-    // Discard any existing exchange context. Effectively we can only have one exchange per ReadClient
-    // at any one time.
-    AbortExistingExchangeContext();
+    ReadRequest::Builder request;
+    err = request.Init(&aWriter);
+    SuccessOrExit(err);
 
+    if (aEventPathParamsListSize != 0 && apEventPathParamsList != nullptr)
     {
-        System::PacketBufferTLVWriter writer;
-        ReadRequest::Builder request;
-
-        msgBuf = System::PacketBufferHandle::New(kMaxSecureSduLengthBytes);
-        VerifyOrExit(!msgBuf.IsNull(), err = CHIP_ERROR_NO_MEMORY);
-
-        writer.Init(std::move(msgBuf));
-
-        err = request.Init(&writer);
-        SuccessOrExit(err);
-
-        if (aEventPathParamsListSize != 0 && apEventPathParamsList != nullptr)
+        EventPathList::Builder & eventPathListBuilder = request.CreateEventPathListBuilder();
+        EventPath::Builder eventPathBuilder           = eventPathListBuilder.CreateEventPathBuilder();
+        for (size_t eventIndex = 0; eventIndex < aEventPathParamsListSize; ++eventIndex)
         {
-            EventPathList::Builder & eventPathListBuilder = request.CreateEventPathListBuilder();
-            EventPath::Builder eventPathBuilder           = eventPathListBuilder.CreateEventPathBuilder();
-            for (size_t eventIndex = 0; eventIndex < aEventPathParamsListSize; ++eventIndex)
-            {
-                EventPathParams eventPath = apEventPathParamsList[eventIndex];
-                eventPathBuilder.NodeId(eventPath.mNodeId)
-                    .EventId(eventPath.mEventId)
-                    .EndpointId(eventPath.mEndpointId)
-                    .ClusterId(eventPath.mClusterId)
-                    .EndOfEventPath();
-                SuccessOrExit(err = eventPathBuilder.GetError());
-            }
+            EventPathParams eventPath = apEventPathParamsList[eventIndex];
```

Comment:
consider using `const &` to help the compiler not needing to keep a on-stack copy of a structure